pyexchange/uniswapv2.py

```diff
@@ -0,0 +1,342 @@
+        """ Calculate amount of given inputs to achieve an exact output amount.
+        
+        Desired amount_out must be less than available liquidity or call will fail.
+
+        Args:
+            amount_out: Desired amount of tokens out.
+            path: List of addresses used to form a path for swap 
+        Returns:
+            A list of uint256 reserve amounts required.
+        """
+        assert (isinstance(amount_out, Wad))
+        assert (isinstance(path, List))
+
+        amounts = self._router_contract.functions.getAmountsIn(amount_out.value, path).call()
+        return list(map(lambda amount: Wad.from_number(Web3.fromWei(amount, 'ether')), amounts))
+
+    def add_liquidity(self, amounts: dict, token_a: Token, token_b: Token) -> Transact:
+        """ Add liquidity to arbitrary token pair.
+
+        Args:
+            amounts: dictionary[Wad, Wad, Wad, Wad]
+            token_a: First token in the pool
+            token_b: Second token in the pool
+        Returns:
+            A :py:class:`pymaker.Transact` instance, which can be used to trigger the transaction.
+        """
+        assert (isinstance(amounts, dict))
+        assert (isinstance(token_a, Token))
+        assert (isinstance(token_b, Token))
+
+        addLiquidityArgs = [
+            token_a.address.address,
+            token_b.address.address,
+            amounts['amount_a_desired'].value,
+            amounts['amount_b_desired'].value,
+            amounts['amount_a_min'].value,
+            amounts['amount_b_min'].value,
+            self.account_address.address,
+            self._deadline()
+        ]
+
+        return Transact(self, self.web3, self.router_abi['abi'], self.router_address, self._router_contract,
+                        'addLiquidity', addLiquidityArgs)
+
+    def add_liquidity_eth(self, amounts: dict, token: Token) -> Transact:
+        """ Add liquidity to token-weth pair.
+
+        Args:
+            amounts: dictionary[Wad, Wad, Wad, Wad]
+            token_a: Token side of the pool
+        Returns:
+            A :py:class:`pymaker.Transact` instance, which can be used to trigger the transaction.
+        """
+        assert (isinstance(amounts, dict))
+        assert (isinstance(token, Token))
+
+        addLiquidityArgs = [
+            token.address.address,
+            amounts['amount_token_desired'].value,
+            amounts['amount_token_min'].value,
+            amounts['amount_eth_min'].value,
+            self.account_address.address,
+            self._deadline()
+        ]
+
+        return Transact(self, self.web3, self.router_abi['abi'], self.router_address, self._router_contract,
+                        'addLiquidityETH', addLiquidityArgs, {'value': amounts['amount_eth_desired'].value})
+
+    # TODO: finish implementing
+    # Enable liquidity to be removed from a pool up to a set limit
+    def permit_removal(self, pair, amount: Wad) -> Transact:
```

Comment:
There's not a lot of value keeping this prototype around.  If there was some implementation behind it, I'd recommend commenting it out.  As is, seems best to remove it.

Reply:
Agreed, I've removed it